### PR TITLE
[LoRA] Add fixed-weight routed LoRA reference path

### PR DESCRIPTION
**Pull request by navyanshmahla**
## Purpose

Foundational implementation for #49705.

This PR adds fixed-weight routed LoRA support for requests of the form:

`base(x) + sum_i w_i * LoRA_i(x)`

The goal is to make vLLM capable of carrying and applying multiple LoRA adapters per token before adding learned/router-based MoE-LoRA routing in follow-up work.

This PR includes:
- `LoRARoutingRequest` for grouping multiple concrete `LoRARequest`s.
- `LoRARouteMapping` for per-token adapter slots and weights.
- Shared route-building helpers for v1 GPU/TPU input paths.
- Routed LoRA metadata conversion in the Punica wrapper.
- Reference routed LoRA execution for linear layers.
- Coverage for replicated, column-parallel, row-parallel, QKV, quantized base layers, and fully sharded LoRA.
- Fail-closed behavior for embeddings/logits, which are not supported in this PR.

This PR intentionally does not add:
- learned/router-based MoE-LoRA gates,
- per-layer dynamic routing,
- optimized routed Punica/Triton kernels,
- embedding/logits routed LoRA support.

Those are planned as follow-up work under #49705.

Duplicate-work check:
- Checked RFC #49705 comments.
- Checked open PRs referencing `49705 in:body`; none found.
- Checked open PRs for routed/MoE LoRA keywords. Related PRs #46772 and #39230 are adjacent work, but neither implements routed mixtures of multiple LoRA adapters per request/token.

AI assistance was used while developing this PR. I reviewed the resulting changes and tests.

## Test Plan

- Unit tests for routed request validation and helper behavior.
- Unit tests for v1 routed LoRA state/mapping construction.
- Reference-output tests comparing routed LoRA layer outputs against explicit PyTorch oracles.
- Coverage includes replicated linear, quantized base linear, QKV, fully sharded column/row/QKV paths.

## Test Result

```bash
.venv/bin/python -m pytest tests/lora/test_lora_routing.py tests/lora/
test_layers_utils.py -q
# 18 passed

.venv/bin/pre-commit run ruff-check --files <changed files>
# Passed

.venv/bin/pre-commit run ruff-format --files <changed files>
# Passed

.venv/bin/python -m py_compile vllm/lora/layers/base_linear.py vllm/lora/
layers/column_parallel_linear.py vllm/lora/layers/row_parallel_linear.py
tests/lora/test_lora_routing.py vllm/lora/routing_utils.py
# Passed

git diff --check origin/main..HEAD
# Passed

No model evals were run because this PR adds infrastructure/reference execution tests and does not change default model behavior unless routed LoRA requests are explicitly used.

### CODE REVIEW

**Comment by navyanshmahla** [comment]
I inspected the `buildkite/intel-ci` failure. The failed job is `Model Runner V2 Examples (Intel)` in build #8321  `examples/features/ automatic_prefix_caching/prefix_caching_offline.py`.

The failing run initializes `facebook/opt-125m` with `enable_lora=None` and the root cause appears to be a native XPU segfault after loading AOT compile artifacts:

`!!!!!!! Segfault encountered !!!!!!!`

This does not appear to exercise the routed LoRA path in this PR. The PR touches `punica_xpu.py` which likely caused the `intel-gpu` CI selection but the failed example does not enable LoRA. Could someone retry `buildkite/ intel-ci` or advise if this is a known XPU flake?
